jkube-kit/generator/webapp/src/main/java/org/eclipse/jkube/generator/webapp/WebAppGenerator.java

```diff
@@ -166,27 +147,65 @@ protected Map<String, String> getEnv(AppServerHandler handler) {
     }
 
     return defaultEnv;
   }
 
-  public static Map<String, String> extractEnvVariables(String envConfigValue) {
-    if (StringUtils.isBlank(envConfigValue)) {
-      return Collections.emptyMap();
-    }
+  private boolean isArtifactSameAsPrevious(File sourceFile) throws IOException, NoSuchAlgorithmException {
+    Path path = sourceFile.getParentFile().toPath();
+    byte[] data = Files.readAllBytes(sourceFile.toPath());
+    byte[] hash = MessageDigest.getInstance("MD5").digest(data);
+    String checksum = new BigInteger(1, hash).toString(16);
+    String previousChecksum = getPreviousChecksum(path);
+    log.info("Checksum for %s is %s", sourceFile.getName(), checksum);
+    log.info("Previous checksum for %s is %s", sourceFile.getName(), previousChecksum);
+    boolean isSameChecksum = checksum.equals(previousChecksum);
+    log.info("Same checksum for %s: %s", sourceFile.getName(), isSameChecksum ? "yes" : "no");
+    saveChecksum(path, checksum);
+    return isSameChecksum;
+  }
 
-    return Pattern.compile("\\n\\s*").splitAsStream(envConfigValue) // "envName1=value1\n   envName2=value2"
-        .map(envNameValue -> envNameValue.split("=")) //
-        .filter(e -> e.length == 2) //
-        .collect(Collectors.toMap(e -> e[0], e -> e[1]));
+  private void saveChecksum(Path path, String checksum) {
+    String fileName = path + File.separator + CHECKSUM_FILENAME;
+    File f = new File(fileName);
+    try (ObjectOutputStream oos = new ObjectOutputStream(Files.newOutputStream(f.toPath()))) {
+      log.info("Saving checksum for %s", path);
+      oos.writeObject(checksum);
+      oos.flush();
+    } catch (IOException e) {
+      log.warn("Failed to save checksum: " + e.getMessage());
+    }
+  }
 
+  private String getPreviousChecksum(Path path) {
+    String fileName = path + File.separator + CHECKSUM_FILENAME;
+    try {
+      File f = new File(fileName);
+      if (f.exists()) {
+        try (ObjectInputStream ois = new ObjectInputStream(Files.newInputStream(f.toPath()))) {
+          return (String) ois.readObject();
+        }
+      }
+    } catch (IOException | ClassNotFoundException e) {
+      log.warn("Failed to read checksum: " + e.getMessage());
+    }
+    return null;
   }
 
   private AssemblyConfiguration createAssembly(AppServerHandler handler) {
     final File sourceFile = Objects.requireNonNull(JKubeProjectUtil.getFinalOutputArtifact(getProject()),
-        "Final output artifact file was not detected. The project may have not been built. HINT: try to compile and package your application prior to running the container image build task."); 
+            "Final output artifact file was not detected. The project may have not been built. HINT: try to compile and package your application prior to running the container image build task.");
+    try {
+      if (isArtifactSameAsPrevious(sourceFile)) {
```

Comment:
`sourceFile` here contains the application jar/war . Do you want to check whether artifact is same by calculating checksum on source directories or target directories?

Reply:
> `sourceFile` here contains the application jar/war . Do you want to check whether artifact is same by calculating checksum on source directories or target directories?

It works for source file, so I don't think we need to change it.